### PR TITLE
feat: add support for batch message handling

### PR DESCRIPTION
**Pull request by frsechet**
See #128 for use case description

### CODE REVIEW

**Comment by frsechet** [comment]
As of https://github.com/bbc/sqs-consumer/pull/129/commits/ccdd3e839bb5eb37229b29f9774e52e9ef8f5a8f, if one message fails, the whole batch is rejected.

My reasoning for implementing it like so is the following:
- the role of the batch handling function is to handle ALL the message. If some fail, the function did not work entirely and throws. This is similar to the following:
```javascript
function applyStuff(array, retries = 0) {
	try {
		array.forEach(doSomethingThatSometimeFails);
	}
	catch(err) {
		if (retries > 5) throw new Error('Max retries exceeded');
		applyStuff(array, retries + 1);
	}
}
```
`doSomethingThatSometimeFails` may be applied to some elements before throwing, and will be applied another time at each retry.
- a SQS message is delivered "at least once", which means that in all cases the application should handle cases where a message is handled twice.

Hence, if you intend to use batch handling, you should probably handle this behaviour in your logic.

I would also find it perfectly acceptable to decide that the handling function should return the successful and errored messages, and filter them out. However, it's a bit more complicated to handle, and it might be a bit too much considering my remarks above.

**TL;DR:**
- either the processing of the whole BATCH is successful, or it is considered failed entirely
- have your code handle retries properly

**Comment by tuan-pham** [comment]
> See #128 for use case description

This will be handy! Do you know when will this be merged to master?

**Comment by wkentdag** [comment]
Super valuable feature. IMO a more intuitive way of processing batch messages than what's currently available.

**Comment by nspragg** [comment]
@frsechet - see https://github.com/bbc/sqs-consumer/issues/128

Are you interested in updating this for Typescript?

**Comment by frsechet** [comment]
Hi @nspragg 

I did most of the migration work, however I don't have a project right now I can test it with. Not a fan of typescript myself, so perhaps you may want to do the last tweaks yourself. I did a bit of refactoring for the version I proposed in September that I couldn't do this afternoon again, as I wasn't expecting the need to rewrite it almost entirely...

You can use this as a start or rewrite the feature, as you prefer. Let me know!

**Comment by nspragg** [comment]
@frsechet - I'll make a few Typescript modifications and get this merged. 

Thanks for your contribution.

**Comment by frsechet** [comment]
Great, thanks! 😃
